File: python/construct-binarytree-from-inorder-postorder-traversal.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Linear time complexity
# due to the use of a dictionary
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        def dfs(left, right, inOrdDic, inorder, postorder):
            if left > right:
                return None
            if left == right:
                postorder.pop()
                return TreeNode(inorder[left])
            
            root_idx = inOrdDic[postorder[-1]]
            root = TreeNode(inorder[root_idx])
            postorder.pop()
            
            root.right = dfs(root_idx+1, right, inOrdDic, inorder, postorder)
            root.left = dfs(left, root_idx-1, inOrdDic, inorder, postorder)
            
            return root
        
        inOrdDic = {el: i for i, el in enumerate(inorder)}
        return dfs(0, len(inorder)-1, inOrdDic, inorder, postorder)

# Quadratic time complexity
# Because of linear searches for the root node in the inorder traversal
# and the slicing of the inorder traversal
class Solution:
    def buildTree2(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        def dfs(inorder, postorder):
            if not inorder:
                return None
            elif len(inorder) == 1:
                postorder.pop()
                return TreeNode(inorder[0])
            
            root_idx = inorder.index(postorder[-1])
            root = TreeNode(inorder[root_idx])
            
            postorder.pop()
            root.right = dfs(inorder[root_idx+1:], postorder)
            root.left = dfs(inorder[:root_idx], postorder)
            
            return root
        
        return dfs(inorder, postorder)

File: go/construct_binarytree_from_inorder_postorder_traversal.go
package leetcode

func buildTree1(inorder []int, postorder []int) *TreeNode {
	postorderIdx := len(inorder) - 1
	return dfs1(0, len(inorder)-1, &postorderIdx, inorder, postorder)
}

func dfs1(left int, right int, postorderIdx *int, inorder []int, postorder []int) *TreeNode {
	if left > right {
		return nil
	}
	if left == right {
		*postorderIdx = (*postorderIdx) - 1
		return &TreeNode{inorder[left], nil, nil}
	}

	rootIdx := findElIdx(left, right, postorder[*postorderIdx], inorder)
	root := TreeNode{inorder[rootIdx], nil, nil}
	*postorderIdx = (*postorderIdx) - 1

	root.Right = dfs1(rootIdx+1, right, postorderIdx, inorder, postorder)
	root.Left = dfs1(left, rootIdx-1, postorderIdx, inorder, postorder)

	return &root

}

func findElIdx(left int, right int, el int, inorder []int) int {
	for i := left; i <= right; i++ {
		if inorder[i] == el {
			return i
		}
	}
	return -1
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}
